src/main/java/org/opensearch/security/dlic/rest/api/SimulationApiAction.java

```
@@ -0,0 +1,411 @@
+                    irr,
+                    indexNameExpressionResolver,
+                    actionPrivileges
+            );
+
+            log.info("Using custom simulation config for user: {}", context.getUser().getName());
+
+            return securityApiDependencies.privilegesEvaluator().evaluate(context);
+
+        } catch (IOException e) {
+            throw new RuntimeException("Error while simulating permissions using custom configuration: ", e);
+        }
+    }
+
+    private ActionPrivileges createActionPrivileges(
+            Map<String, Object> proposedRoles,
+            Map<String, Object> proposedActionGroups,
+            Settings settings
+    ) throws IOException {
+
+        SecurityDynamicConfiguration<RoleV7> rolesConfig = SecurityDynamicConfiguration.fromMap(proposedRoles, CType.ROLES);
+
+        SecurityDynamicConfiguration<ActionGroupsV7> actionGroupsConfig = SecurityDynamicConfiguration.fromMap(proposedActionGroups, CType.ACTIONGROUPS);
+        FlattenedActionGroups flattenedActionGroups = new FlattenedActionGroups(actionGroupsConfig.withStaticConfig());
+
+        return new RoleBasedActionPrivileges(
+                rolesConfig.withStaticConfig(),
+                flattenedActionGroups,
+                settings
+        );
+    }
+
+    @SuppressWarnings("unchecked")
+    private PrivilegesEvaluationContext createPrivilegesEvaluationContext(
+            Map<String, Object> proposedBody,
+            ThreadPool threadPool,
+            ClusterService clusterService,
+            IndexResolverReplacer irr,
+            IndexNameExpressionResolver indexNameExpressionResolver,
+            ActionPrivileges actionPrivileges )
+    {
+        ThreadContext threadContext = threadPool.getThreadContext();
+        String action = (String) proposedBody.get(ACTION);
+        if (action == null || action.isEmpty()) {
+            throw new IllegalArgumentException("Missing action in request body");
+        }
```

These all checks must be done before processing request. We must always fail fast if input is not as expected

---

updated